### ISSUE TITLE
Remove "probably" warnings

### ISSUE THREAD

**Issue by StabbyCutyou** [golang/lint#234]
Specifically around replacing something like

`can probably use var x []int` instead of `make([]int,0)`

The reason why I feel this should be removed has been stated in your docs: "We will not be adding pragmas or other knobs to suppress specific warnings". A warning that says "you can probably do this" feels much more like someones opinion rather than an actionable warning.

I understand that in some circles, the `var x []int` approach is preferable, but as go allows the wiggle room to declare a slice of 0 in more than one way, then the linter itself (where it will never have a way to suppress such a warning) should not attempt to enforce one way over the other - especially when that way is only applicable in one particular situation, and is not universally applicable (ie, you cannot make the same suggestion for a slice with a size >= 1).

Thanks

**Comment by bgreezy** [comment]
I agree but will add the following suggestion: Could it be made so that 'probably' warnings still exist but are optional according to a suppression flag?

**Comment by StabbyCutyou** [comment]
The author stated they have no intention for such flags, which I can
respect.

But in such a case, I think the tool should only warn on concrete
actionable items.

**Comment by dsnet** [comment]
I'm trying to understand the use-case here better. Are you using `make([]int, 0)` for semantic reasons? That is, `a := make([]int, 0)` creates a non-nil slice, while `var a []int` creates a nil slice.

If it's for purely stylistic reasons, isn't that an argument in favor of making this _always_ actionable (that is dropping the "can probably ") and suggest that people always use `var a []int` instead? Part of the point of having a style-checker is to help enforce uniformity in how we write code.

**Comment by StabbyCutyou** [comment]
So, let me try and illustrate why I think this leads to inconsistent-looking code by giving you a straw example.

I fully understand that you'd likely declare a bunch of slices like this all at once, and in practice what you'd see would be these declarations scattered around different places in code for different purposes - this is solely to illustrate a point about consistent-looking code.

Example 1
```
var s1 []int
s2 := make([]int,5)
s3 := make([]string, 1)
```

In example one, you're declaring slices two different ways entirely. The code is *inconsistent* both in how it's read, and how it's written.

Example 2
```
s1 := make([]int,0)
s2 := make([]int,5)
s3 := make([]string, 1)
```

Each slice is declared using the same syntax. The code is *consistent* in how it's read and written.

Ultimately, some people will simply prefer Example 1 because they enjoy the tricks go lets you do with nil slices, but some people will not. I am not saying either approach is right, or wrong, but rather that the 2nd approach is a more *consistent* way to write code.

Given that the tool (rightfully) refuses to budge on adding knobs or dials to control the level of complaints, my point is simply it should not take a personal preference on issues where personal preference can exist (a rarity in go, if we're being honest), and should not nag about "Probably" changing something.

**Comment by cyouden** [comment]
@dsnet Here's a use case where it matters and this check bit me: when converting to JSON, `nil` maps to JSON `null` but `make([]int, 0)` maps to an empty array in JSON (i.e. `[]`), and there are cases where that is preferable (i.e. the consumers of my service don't need to make a null check if I can just return an empty array).

It was possible to get around by adding a check to see if the slice was nil and then assigning a non-nil empty slice (this raises a new question though: if it's bad at initialization, why isn't it bad elsewhere?), but that check is three lines of unnecessary code to do the same thing.

But the short of it is: by following golint's suggestion, I actually _created_ a bug. That seems counterproductive for a linter.

**Comment by jcsalem** [comment]
I've had this same issue with JSON conversion as well as some related problems with mocking tools.

`var x []int` and `x := make([]int,0)` don't have the same effect and golint should not make that suggestion at all.

Alternatively, `x := []int{}` and `x := make([]int,0)` do have the same effect and perhaps golint could make a suggestion there.

**Comment by StabbyCutyou** [comment]
Thank you so much for accepting this!

**Comment by dlsniper** [comment]
I'd like to point out that the declarations have different effects: 
```go
package main

import "fmt"

func main() {
	a := []int{}
	b := make([]int, 0)
	var c []int
	fmt.Printf("a: %p len: %d cap: %d\n", a, len(a), cap(a))
	fmt.Printf("b: %p len: %d cap: %d\n", b, len(b), cap(b))
	fmt.Printf("c: %p len: %d cap: %d\n", c, len(c), cap(c))
}
```

Output:
```
a: 0x196a9c len: 0 cap: 0
b: 0x196a9c len: 0 cap: 0
c: 0x0 len: 0 cap: 0
```

https://play.golang.org/p/tJjMp8Ixtq

I'd rather have this back but explained why ` var x []int ` should be preferred over ` x := []int{} ` or ` x := make([]int, 0) ` and when.

**Comment by ScottMansfield** [comment]
@dlsniper they have different implementations but have the same behavior at a language level.

**Comment by StabbyCutyou** [comment]
The point is that in some cases, the existing warning could produce incorrect code. This is a problem if you make, for example, passing go-lint a build requirement.

Leaving it up to the author to determine the right thing at the right time is a better solution.

**Comment by dsnet** [comment]
That's pretty much what the original check did by suggesting that you "can probably use" one form or the other. The issue is that a one-line lint message isn't going to do justice in explaining how to distinguish between the two usages. Secondly, I'm not sure if it's lint's responsibility to teach this sort of thing.

**Comment by cyouden** [comment]
@dlsniper My two cents: Leaving it in is an issue since there's no way to tell it that you meant to do that in that specific instance, so then you're training developers to ignore messages from the linter. Eventually, if there are enough ignored messages, the real ones get lost in the noise, which defeats the purpose. Since, as I understand it, there are no plans to make golint configurable (like comments to tell it to ignore specific lines), I think the correct solution is to only include things that are _always_ wrong, rather than things that are only _sometimes_ wrong.

**Comment by shurcooL** [comment]
@StabbyCutyou, since it wasn't mentioned in this thread so far, I'll add a quick note for completeness. You mentioned these two alternatives:

```Go
var s1 []int
s2 := make([]int, 5)
s3 := make([]string, 1)
```

```Go
s1 := make([]int, 0)
s2 := make([]int, 5)
s3 := make([]string, 1)
```

The second being more consistent in how it's written. Here's a 3rd alternative for consideration, one that is a _little_ more consistent in how it's written while still giving a `nil` slice:

```Go
s1 := []int(nil)
s2 := make([]int, 5)
s3 := make([]string, 1)
```

It might be helpful for those (very rare) situations where you care whether the empty slice is nil or non-nil.
